File: hdf-converters/shared/go/cvss.go
package shared

import (
	"strings"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	hdfutil "github.com/mitre/hdf-libs/hdf-utilities/go/v3"
)

// CvssVersionFromVector returns the schema CVSS Version enum for a vector
// string's prefix (CVSS:2.0/, CVSS:3.0/, CVSS:3.1/, CVSS:4.0/). An optional
// fallback sets the version returned when the vector carries no recognized
// prefix (e.g. historical Nessus output with no prefix passes 3.0); when
// omitted the fallback is 3.1, the version modern scanners emit most often.
func CvssVersionFromVector(vector string, fallback ...hdf.Version) hdf.Version {
	def := hdf.The31
	if len(fallback) > 0 {
		def = fallback[0]
	}
	switch {
	case strings.HasPrefix(vector, "CVSS:2.0/"):
		return hdf.The20
	case strings.HasPrefix(vector, "CVSS:3.0/"):
		return hdf.The30
	case strings.HasPrefix(vector, "CVSS:3.1/"):
		return hdf.The31
	case strings.HasPrefix(vector, "CVSS:4.0/"):
		return hdf.The40
	default:
		return def
	}
}

// CvssVersionFromString maps a bare CVSS version number ("2.0", "3.0", "3.1",
// "4.0") — as emitted in a structured version field rather than a vector
// prefix — to the schema Version enum. Unrecognized values default to "3.1".
func CvssVersionFromString(v string) hdf.Version {
	switch v {
	case "2.0":
		return hdf.The20
	case "3.0":
		return hdf.The30
	case "4.0":
		return hdf.The40
	default:
		return hdf.The31
	}
}

// CvssSeverityFromScore maps a CVSS base/computed score to the schema
// CVSSSeverity enum via the shared band thresholds in hdfutil. Scores below the
// low threshold (and non-finite inputs, per hdfutil) map to the "none" band.
func CvssSeverityFromScore(score float64) hdf.CVSSSeverity {
	switch hdfutil.CvssScoreToSeverity(score) {
	case "critical":
		return hdf.CVSSSeverityCritical
	case "high":
		return hdf.CVSSSeverityHigh
	case "medium":
		return hdf.CVSSSeverityMedium
	case "low":
		return hdf.CVSSSeverityLow
	default:
		return hdf.None
	}
}

// CvssInput carries the base-metric fields a converter provides for a single
// Cvss entry. Empty/nil fields are omitted from the assembled entry, so each
// converter can produce exactly the shape it produces today (some set a
// BaseVector and Source, some do not).
type CvssInput struct {
	Version    hdf.Version
	BaseScore  *float64
	BaseVector string
	Source     string
}

// BuildCvss assembles an hdf.Cvss from the base-metric fields a converter
// provides. Version is always set. BaseScore (and its derived BaseSeverity) is
// set only when non-nil; BaseVector and Source are set only when non-empty.
// Consumers needing threat/environmental/computed enrichment set those fields
// on the returned value themselves.
func BuildCvss(in CvssInput) hdf.Cvss {
	cv := hdf.Cvss{Version: in.Version}
	if in.BaseScore != nil {
		score := *in.BaseScore
		cv.BaseScore = &score
		sev := CvssSeverityFromScore(score)
		cv.BaseSeverity = &sev
	}
	if in.BaseVector != "" {
		bv := in.BaseVector
		cv.BaseVector = &bv
	}
	if in.Source != "" {
		src := in.Source
		cv.Source = &src
	}
	return cv
}

File: hdf-converters/shared/typescript/cvss.ts
import { cvssScoreToSeverity } from '@mitre/hdf-utilities';
import type { Cvss } from '@mitre/hdf-schema';
import { CVSSSeverity, Version as CvssVersion } from '@mitre/hdf-schema';

/**
 * Returns the schema CVSS Version enum for a vector-string prefix (CVSS:2.0/,
 * CVSS:3.0/, CVSS:3.1/, CVSS:4.0/). `defaultVersion` sets the version returned
 * when the vector is absent or carries no recognized prefix (e.g. historical
 * Nessus output with no prefix passes 3.0); it defaults to 3.1, the version
 * modern scanners emit most often.
 */
export function cvssVersionFromVector(
  vector: string | undefined,
  defaultVersion: CvssVersion = CvssVersion.The31
): CvssVersion {
  if (!vector) return defaultVersion;
  if (vector.startsWith('CVSS:2.0/')) return CvssVersion.The20;
  if (vector.startsWith('CVSS:3.0/')) return CvssVersion.The30;
  if (vector.startsWith('CVSS:3.1/')) return CvssVersion.The31;
  if (vector.startsWith('CVSS:4.0/')) return CvssVersion.The40;
  return defaultVersion;
}

/**
 * Maps a bare CVSS version number ("2.0"/"3.0"/"3.1"/"4.0") — as emitted in a
 * structured version field rather than a vector prefix — to the schema Version
 * enum. Unrecognized values default to 3.1.
 */
export function cvssVersionFromString(v: string | undefined): CvssVersion {
  switch (v) {
    case '2.0': return CvssVersion.The20;
    case '3.0': return CvssVersion.The30;
    case '4.0': return CvssVersion.The40;
    default: return CvssVersion.The31;
  }
}

/**
 * Maps a CVSS base/computed score to the schema CVSSSeverity enum via the shared
 * band thresholds in cvssScoreToSeverity. Scores below the low threshold (and
 * non-finite inputs, per cvssScoreToSeverity) map to the "none" band.
 */
export function cvssSeverityFromScore(score: number): CVSSSeverity {
  switch (cvssScoreToSeverity(score)) {
    case 'critical': return CVSSSeverity.Critical;
    case 'high': return CVSSSeverity.High;
    case 'medium': return CVSSSeverity.Medium;
    case 'low': return CVSSSeverity.Low;
    default: return CVSSSeverity.None;
  }
}

/**
 * Base-metric fields a converter provides for a single Cvss entry. Empty/undefined
 * fields are omitted from the assembled entry, so each converter can produce
 * exactly the shape it produces today (some set a baseVector and source, some do not).
 */
export interface CvssInput {
  version: CvssVersion;
  baseScore?: number | null;
  baseVector?: string | null;
  source?: string | null;
}

/**
 * Assembles a Cvss from the base-metric fields a converter provides. version is
 * always set. baseScore (and its derived baseSeverity) is set only when a finite
 * number is supplied; baseVector and source are set only when non-empty. Consumers
 * needing threat/environmental/computed enrichment set those fields on the result.
 */
export function buildCvss(input: CvssInput): Cvss {
  const cv: Cvss = {version: input.version};
  if (typeof input.baseScore === 'number' && Number.isFinite(input.baseScore)) {
    cv.baseScore = input.baseScore;
    cv.baseSeverity = cvssSeverityFromScore(input.baseScore);
  }
  if (input.baseVector) cv.baseVector = input.baseVector;
  if (input.source) cv.source = input.source;
  return cv;
}
